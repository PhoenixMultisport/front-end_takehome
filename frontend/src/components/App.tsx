import * as React from 'react';
import { useEffect, useState } from 'react';
import axios, { AxiosResponse } from 'axios';
import PractitionerFeedListItem from './PractitionerFeedListItem';
import { HapiPractitionerEntryItem, HapiPractitionersResponse } from '../types/Hapi.types';
import { getPractitionerEmail } from '../utils/helpers';

function App() {
  const [practitioners, setPractitioners] = useState<HapiPractitionerEntryItem[]>([]);

  useEffect(() => {
    axios.get<HapiPractitionersResponse>('http://hapi.fhir.org/baseDstu3/Practitioner')
      .then((res: AxiosResponse<HapiPractitionersResponse>) => setPractitioners(res.data.entry))
      .catch((err) => console.error(err));
  }, [])

  return (
    <div className="text-center">
      <header>
      </header>
      <main className="bg-sky-900 text-white min-h-screen flex flex-col items-center justify-center content-center">
        <ul>
          {
            practitioners.map((p: HapiPractitionerEntryItem, i: number) => {
              const practitionerEmail: string = getPractitionerEmail(p.resource.telecom);

              return (
                <PractitionerFeedListItem
                  key={i}
                  itemKey={p.resource.id}
                  resourceType={p.resource.resourceType}
                  email={practitionerEmail}
                  url={p.fullUrl}
                />
              )
            })
          }
        </ul>
      </main>
    </div>
  );
}

export default App;
