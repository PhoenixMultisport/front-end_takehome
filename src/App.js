import React from 'react';
import { Provider } from 'react-redux';
import { ToastContainer } from 'react-toastify';
import { store } from './redux/store';
import AppRouter from './router';

import 'react-toastify/dist/ReactToastify.min.css';
import 'antd/dist/antd.min.css';
import './App.css';

function App() {
  return (
    <Provider store={store}>
      <ToastContainer />
      <AppRouter />
    </Provider>
  );
}

export default App;
