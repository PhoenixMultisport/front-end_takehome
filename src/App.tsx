import React, { useState, useEffect } from 'react';
import ApiService from './_api/api.fhir.js';
import Practitioner from './Components//Practitioner/Practitioner';
import { Col, Row } from 'antd';
import helper from './Helpers/helper';
import './Models/Practitioner';
import './_styles/App.css';

function App() {

  //state
  const [practitioners, setPractitioners] = useState<Practitioner[]>([]);
  const [selectedPractitioner, setSelectedPractitioner] = useState<Practitioner | null>(null);

  //ui helpers
  const [loading, setLoading] = useState(false);


  // ****************************** //
  // ************ API ************* //
  // ****************************** //

  const initialLoadPractitioners = async() => {
    try {
      //throw up loading icon
      setLoading(true);

      //get the data
      const practitionerList = await ApiService.httpGet();

      //set state
      setPractitioners(practitionerList.entry);

    } catch (err: any) {
      console.error(err.message);
    } finally {
      setLoading(false);
    }
  };


  // *********************************** //
  // ************ Handlers ************* //
  // *********************************** //

  const handleSelectPractitioner = async(pracitionerId: string) => {
    try {
      setLoading(true);
      const practitioner = await ApiService.httpGet(pracitionerId);

      if (practitioner) {
        setSelectedPractitioner(practitioner);
      }

      setLoading(false);
    } catch (err: any) {
      console.error(err.message);
    } finally {
      setLoading(false);
    }
  };

  const handleClearPractitioner = () => {
    setSelectedPractitioner(null);
  };


  // *************************************** //
  // ************ Default Load ************* //
  // *************************************** //

  useEffect(() => {
    //set empty dependency array so it only loads once on mount
    initialLoadPractitioners();
  }, []);


  if (loading) {
    return <h1>Loading...</h1>
  }

  return (
    <div className="App">
      <Row>
        {!selectedPractitioner && (
          <>
            <Col span={12} offset={6}>
                <h1>Practitioners</h1>
              </Col>
              <Col span={12} offset={6}>
                {practitioners && practitioners.length > 0 &&
                  <ul>
                    {practitioners.map((practitioner, index) => {

                      const nameValue = helper.assignFieldNameValue(practitioner.resource);
                      const emailValue = helper.assignFieldEmailValue(practitioner.resource);

                      return (
                        <li key={index} onClick={() => handleSelectPractitioner(practitioner.resource.id)}>
                          {nameValue ? nameValue : "Name unavailable"} / {emailValue ? emailValue : "Email unavailable"}
                        </li>
                      );
                    })}
                  </ul>
                }
              </Col>
          </>
        )}
        {selectedPractitioner && (
          <Col span={24}>
            <Practitioner
              selectedPractitioner={selectedPractitioner}
              handleClearPractitioner={handleClearPractitioner}
            />
          </Col>
        )}
      </Row>
    </div>
  );
}

export default App;
