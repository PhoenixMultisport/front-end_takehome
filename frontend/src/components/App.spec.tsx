import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { describe } from '@jest/globals';
import App from './App';
import { Axios } from 'axios';

let documentBody: RenderResult;

describe('<App />', () => {
  beforeEach(async () => {
    await process.nextTick(() => {});

    documentBody = render(
      <App/>
    );

    jest.spyOn(Axios.prototype, 'get').mockResolvedValue([]);
  });

  it('renders without crashing', () => {
    expect(documentBody).toBeTruthy();
  });

  it('matches snapshot', () => {
    const { baseElement } = documentBody;

    expect(baseElement).toMatchSnapshot();
  });
})
