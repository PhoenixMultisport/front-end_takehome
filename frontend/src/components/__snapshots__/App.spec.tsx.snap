// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<body>
  <div>
    <div
      class="text-center"
    >
      <header />
      <main
        class="bg-sky-900 text-white min-h-screen flex flex-col items-center justify-center content-center"
      >
        <ul />
      </main>
    </div>
  </div>
</body>
`;
